With docker==2.2.0, the `secrets` tests crash:
```
TASK [secrets] ***********************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/home/felix/.ansible/tmp/ansible-tmp-1551824244.4349453-189620470344759/AnsiballZ_docker_swarm_service.py:18: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses\n  import imp\nTraceback (most recent call last):\n  File \"/home/felix/.ansible/tmp/ansible-tmp-1551824244.4349453-189620470344759/AnsiballZ_docker_swarm_service.py\", line 114, in <module>\n    _ansiballz_main()\n  File \"/home/felix/.ansible/tmp/ansible-tmp-1551824244.4349453-189620470344759/AnsiballZ_docker_swarm_service.py\", line 106, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/felix/.ansible/tmp/ansible-tmp-1551824244.4349453-189620470344759/AnsiballZ_docker_swarm_service.py\", line 49, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/usr/lib/python3.7/imp.py\", line 234, in load_module\n    return load_source(name, filename, file)\n  File \"/usr/lib/python3.7/imp.py\", line 169, in load_source\n    module = _exec(spec, sys.modules[name])\n  File \"<frozen importlib._bootstrap>\", line 630, in _exec\n  File \"<frozen importlib._bootstrap_external>\", line 728, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n  File \"/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py\", line 2302, in <module>\n  File \"/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py\", line 2285, in main\n  File \"/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py\", line 2046, in run_safe\n  File \"/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py\", line 2036, in run\n  File \"/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py\", line 1918, in create_service\n  File \"/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py\", line 1706, in build_docker_service\n  File \"/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py\", line 1532, in build_container_spec\n  File \"/usr/lib/python3.7/site-packages/docker/utils/decorators.py\", line 19, in wrapped\n    'Resource ID was not provided'\ndocker.errors.NullResource: Resource ID was not provided\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
...ignoring
```
The relevant part of the stacktrace:
```
  File "/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py", line 2302, in <module>
  File "/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py", line 2285, in main
  File "/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py", line 2046, in run_safe
  File "/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py", line 2036, in run
  File "/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py", line 1918, in create_service
  File "/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py", line 1706, in build_docker_service
  File "/tmp/ansible_docker_swarm_service_payload_etwa8cly/__main__.py", line 1532, in build_container_spec
  File "/usr/lib/python3.7/site-packages/docker/utils/decorators.py", line 19, in wrapped
    'Resource ID was not provided'
docker.errors.NullResource: Resource ID was not provided
```

--- a/lib/ansible/modules/cloud/docker/docker_swarm_service.py
+++ b/lib/ansible/modules/cloud/docker/docker_swarm_service.py
@@ -658,7 +658,7 @@ extends_documentation_fragment:
   - docker
   - docker.docker_py_2_documentation
 requirements:
-  - "docker >= 2.0"
+  - "docker >= 2.0.2"
   - "Docker API >= 1.24"
 notes:
   - "Images will only resolve to the latest digest when using Docker API >= 1.30 and docker-py >= 3.2.0.
@@ -2269,7 +2269,7 @@ def main():
         argument_spec=argument_spec,
         required_if=required_if,
         supports_check_mode=True,
-        min_docker_version='2.0.0',
+        min_docker_version='2.0.2',
         min_docker_api_version='1.24',
         option_minimal_versions=option_minimal_versions,
     )

--- a/lib/ansible/modules/cloud/docker/docker_swarm_service.py
+++ b/lib/ansible/modules/cloud/docker/docker_swarm_service.py
@@ -1943,7 +1943,10 @@ class DockerServiceManager(object):
 
     def can_update_networks(self):
         # Before Docker API 1.29 adding/removing networks was not supported
-        return self.client.docker_api_version >= LooseVersion('1.29')
+        return (
+            self.client.docker_api_version >= LooseVersion('1.29') and
+            self.client.docker_py_version >= LooseVersion('2.7')
+        )
 
     def run(self):
         self.diff_tracker = DifferenceTracker()

--- a/test/integration/targets/docker_swarm_service/tasks/tests/options.yml
+++ b/test/integration/targets/docker_swarm_service/tasks/tests/options.yml
@@ -191,12 +191,13 @@
     - configs_3 is changed
     - configs_4 is changed
     - configs_5 is not changed
-  when: docker_api_version is version('1.30', '>=')
+  when: docker_api_version is version('1.30', '>=') and docker_py_version is version('2.5.0', '>=')
+
 - assert:
     that:
     - configs_1 is failed
     - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.30') in configs_1.msg"
-  when: docker_api_version is version('1.30', '<')
+  when: docker_api_version is version('1.30', '<') or docker_py_version is version('2.5.0', '<')
 
 ####################################################################
 ## command #########################################################
@@ -420,12 +421,12 @@
       - dns_4 is changed
       - dns_5 is changed
       - dns_6 is not changed
-  when: docker_api_version is version('1.25', '>=')
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('2.6.0', '>=')
 - assert:
     that:
     - dns_1 is failed
     - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in dns_1.msg"
-  when: docker_api_version is version('1.25', '<')
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('2.6.0', '<')
 
 ####################################################################
 ## dns_options #####################################################
@@ -495,12 +496,12 @@
       - dns_options_3 is changed
       - dns_options_4 is changed
       - dns_options_5 is not changed
-  when: docker_api_version is version('1.25', '>=')
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('2.6.0', '>=')
 - assert:
     that:
     - dns_options_1 is failed
     - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in dns_options_1.msg"
-  when: docker_api_version is version('1.25', '<')
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('2.6.0', '<')
 
 ####################################################################
 ## dns_search ######################################################
@@ -576,12 +577,12 @@
       - dns_search_4 is changed
       - dns_search_5 is changed
       - dns_search_6 is not changed
-  when: docker_api_version is version('1.25', '>=')
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('2.6.0', '>=')
 - assert:
     that:
     - dns_search_1 is failed
     - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in dns_search_1.msg"
-  when: docker_api_version is version('1.25', '<')
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('2.6.0', '<')
 
 ####################################################################
 ## endpoint_mode ###################################################
@@ -622,6 +623,12 @@
       - endpoint_mode_1 is changed
       - endpoint_mode_2 is not changed
       - endpoint_mode_3 is changed
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('3.0.0', '>=')
+- assert:
+    that:
+    - endpoint_mode_1 is failed
+    - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in endpoint_mode_1.msg"
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('3.0.0', '<')
 
 ####################################################################
 ## env #############################################################
@@ -821,12 +828,12 @@
     that:
       - force_update_1 is changed
       - force_update_2 is changed
-  when: docker_api_version is version('1.25', '>=')
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('2.1.0', '>=')
 - assert:
     that:
     - force_update_1 is failed
     - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in force_update_1.msg"
-  when: docker_api_version is version('1.25', '<')
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('2.1.0', '<')
 
 ####################################################################
 ## groups ##########################################################
@@ -890,6 +897,12 @@
     - groups_3 is changed
     - groups_4 is changed
     - groups_5 is not changed
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('2.6.0', '>=')
+- assert:
+    that:
+    - groups_1 is failed
+    - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in groups_1.msg"
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('2.6.0', '<')
 
 ####################################################################
 ## healthcheck #####################################################
@@ -1011,12 +1024,12 @@
     - healthcheck_7 is not changed
     - healthcheck_8 is changed
     - healthcheck_9 is not changed
-  when: docker_py_version is version('2.4.0', '>=')
+  when: docker_api_version is version('1.25', '>=')
 - assert:
     that:
     - healthcheck_1 is failed
-    - "('version is ' ~ docker_py_version ~'. Minimum version required is 2.4.0') in healthcheck_1.msg"
-  when: docker_py_version is version('2.4.0', '<')
+    - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in healthcheck_1.msg"
+  when: docker_api_version is version('1.25', '<')
 
 ###################################################################
 ## hostname #######################################################
@@ -1057,12 +1070,12 @@
       - hostname_1 is changed
       - hostname_2 is not changed
       - hostname_3 is changed
-  when: docker_api_version is version('1.25', '>=')
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('2.2.0', '>=')
 - assert:
     that:
     - hostname_1 is failed
     - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in hostname_1.msg"
-  when: docker_api_version is version('1.25', '<')
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('2.2.0', '<')
 
 ###################################################################
 ## hosts ##########################################################
@@ -1514,6 +1527,12 @@
     - stop_signal_1 is changed
     - stop_signal_2 is not changed
     - stop_signal_3 is changed
+  when: docker_api_version is version('1.28', '>=') and docker_py_version is version('2.6.0', '>=')
+- assert:
+    that:
+    - stop_signal_1 is failed
+    - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.28') in stop_signal_1.msg"
+  when: docker_api_version is version('1.28', '<') or docker_py_version is version('2.6.0', '<')
 
 ####################################################################
 ## publish #########################################################
@@ -1623,12 +1642,12 @@
       - publish_5 is not changed
       - publish_6 is changed
       - publish_7 is not changed
-  when: docker_api_version is version('1.25', '>=')
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('3.0.0', '>=')
 - assert:
     that:
     - publish_1 is failed
     - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in publish_1.msg"
-  when: docker_api_version is version('1.25', '<')
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('3.0.0', '<')
 
 ###################################################################
 ## replicas #######################################################
@@ -1779,12 +1798,12 @@
       - secrets_3 is changed
       - secrets_4 is changed
       - secrets_5 is not changed
-  when: docker_api_version is version('1.25', '>=')
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('2.1.0', '>=')
 - assert:
     that:
     - secrets_1 is failed
     - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in secrets_1.msg"
-  when: docker_api_version is version('1.25', '<')
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('2.1.0', '<')
 
 ###################################################################
 # tty #############################################################
@@ -1825,12 +1844,12 @@
       - tty_1 is changed
       - tty_2 is not changed
       - tty_3 is changed
-  when: docker_api_version is version('1.25', '>=')
+  when: docker_api_version is version('1.25', '>=') and docker_py_version is version('2.4.0', '>=')
 - assert:
     that:
     - tty_1 is failed
     - "('version is ' ~ docker_api_version ~'. Minimum version required is 1.25') in tty_1.msg"
-  when: docker_api_version is version('1.25', '<')
+  when: docker_api_version is version('1.25', '<') or docker_py_version is version('2.4.0', '<')
 
 ###################################################################
 ## user ###########################################################

